Improve cancellability in TransportTasksAction

DaveCTurner opened the pull request:
Each `TransportTasksAction` fans-out to multiple nodes, accumulates responses and retains them until all the nodes have responded, and then converts the responses into a final result.

Similarly to #92987 and #93484, we should accumulate the responses in a structure that doesn't require so much copying later on, and should drop the received responses if the task is cancelled while some nodes' responses are still pending.

DaveCTurner committed:
Improve cancellability in TransportTasksAction

Each `TransportTasksAction` fans-out to multiple nodes, accumulates
responses and retains them until all the nodes have responded, and then
converts the responses into a final result.

Similarly to #92987 and #93484, we should accumulate the responses in a
structure that doesn't require so much copying later on, and should drop
the received responses if the task is cancelled while some nodes'
responses are still pending.

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticsearchmachine commented:
Hi @DaveCTurner, I've created a changelog YAML for you.

DaveCTurner committed:
Update docs/changelog/96279.yaml

DaveCTurner committed:
Must handle exceptions from newResponse too

DaveCTurner committed:
Merge branch 'main' into 2023-05-23-TransportTasksAction-cancellability

arteam commented on server/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/TransportTasksActionTests.java:
```diff
@@ -674,6 +680,152 @@ protected void taskOperation(
+
+        reachabilityChecker.ensureUnreachable();
+
+        while (true) {
```
Would it make sense to use ` while (taskResponseListeners.peek() != null)` here instead of manually breaking out of the loop?

DaveCTurner replied:
That'd work, but I'd rather just do a single read, avoiding having to reason about the relationship between `peek()` and `poll()`.

arteam commented on server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:
```diff
@@ -85,66 +81,164 @@ protected TransportTasksAction(
+                            return;
+                        }
+
+                        logger.debug(Strings.format("failed to execute on node [{}]", nodeId), e);
```
I was wondering if can perform formatting lazily by using a supplier : `logger.debug(() -> Strings.format(`

DaveCTurner replied:
Yes; it's not very common to get here, but I'll do that.

arteam commented on server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:
```diff
@@ -85,66 +81,164 @@ protected TransportTasksAction(
+                    @Override
+                    public void onResponse(NodeTasksResponse nodeResponse) {
+                        synchronized (taskResponses) {
+                            taskResponses.addAll(nodeResponse.results);
```
I've run some tests and have seen a lot of cases where `nodeResponse.results` is empty. Would it make sense to perform an `isEmpty` check on `nodeResponse.results` before acquiring the lock on `taskResponses`?

arteam commented on server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:
```diff
@@ -85,66 +81,164 @@ protected TransportTasksAction(
-    private void nodeOperation(CancellableTask task, NodeTaskRequest nodeTaskRequest, ActionListener<NodeTasksResponse> listener) {
-        TasksRequest request = nodeTaskRequest.tasksRequest;
-        processTasks(request, ActionListener.wrap(tasks -> nodeOperation(task, listener, request, tasks), listener::onFailure));
+        final var taskResponses = new ArrayList<TaskResponse>();
```
Does it make sense to pre-size `taskResponses` based on the amount of the nodes from which we collect responses?

DaveCTurner replied:
I don't think the number of nodes would be a particularly useful estimate for this - as you say, many nodes return nothing, but sometimes they will return thousands of results. These are pretty low-throughput APIs so I think the default sizing is ok.

arteam reviewed (approved):
LGTM! I've left a few cosmeticcomments

DaveCTurner committed:
Merge branch 'main' into 2023-05-23-TransportTasksAction-cancellability

DaveCTurner committed:
Lazy log

DaveCTurner committed:
Fast-path for empty responses

elasticmachine committed:
Merge branch 'main' into 2023-05-23-TransportTasksAction-cancellability

henningandersen reviewed (approved):
LGTM.

Slightly complex (but good), I wonder if a comment or two on how the result lists are captured locally and how cancelling ensures it is freed could make sense?

DaveCTurner commented:
Yeah it is a bit convoluted isn't it? I extracted a utility (and added more commentary) in https://github.com/elastic/elasticsearch/pull/96373, and will update this PR to use it once that's merged.

DaveCTurner committed:
Merge branch 'main' into 2023-05-23-TransportTasksAction-cancellability

DaveCTurner committed:
Use CancellableFanOut

DaveCTurner committed:
Merge branch 'main' into 2023-05-23-TransportTasksAction-cancellability
